grammars/silver/compiler/extension/records/Records.sv.md

```diff
@@ -0,0 +1,370 @@
+this::FieldNames0 ::= fieldNames::FieldNames1  ','
+{ forwards to fieldNames0Nonempty(fieldNames, location=this.location); }
+
+concrete production fieldNames1One
+this::FieldNames1 ::= name::IdLower_t
+{
+  this.unparse = name.lexeme;
+  this.names = [name.lexeme];
+}
+
+concrete production fieldNames1Cons
+this::FieldNames1 ::= init::FieldNames1  ','  last::IdLower_t
+{
+  this.unparse = s"${init.unparse}, ${last.lexeme}";
+  this.names = init.names ++ [last.lexeme];
+}
+```
+
+To make `withoutDeclaring` flow up through the nonterminal modifier lists, we need to add aspects for those nonterminals.
+
+```silver
+aspect production nonterminalModifiersNone
+this::NonterminalModifiers ::=
+{ this.withoutDeclaring = []; }
+
+aspect production nonterminalModifierSome
+this::NonterminalModifiers ::= modifiers::NonterminalModifierList
+{ this.withoutDeclaring = modifiers.withoutDeclaring; }
+
+aspect production nonterminalModifierSingle
+this::NonterminalModifierList ::= modifier::NonterminalModifier
+{ this.withoutDeclaring = modifier.withoutDeclaring; }
+
+aspect production nonterminalModifiersCons
+this::NonterminalModifierList ::= head::NonterminalModifier ',' tail::NonterminalModifierList
+{ this.withoutDeclaring = head.withoutDeclaring ++ tail.withoutDeclaring; }
+```
+
+Once the `withoutDeclaring` values have flowed up through the modifiers, we push them back down to the fields.
+
+```silver
+inherited attribute withoutDeclaringInh::[String];
+
+aspect production recordDecl
+this::AGDcl ::= quals::NTDeclQualifiers  'record'  name::IdUpper_t
+  modifiers::NonterminalModifiers  '{'  fields::RecordFields0  '}'  ';'
+{
+  fields.withoutDeclaringInh = modifiers.withoutDeclaring;
+}
+```
+
+### Fields
+
+Record fields are pretty straightforward, just a pair of a name and a type.
+They define an `AGDcl` they correspond to too, for the particular nonterminal they should occur on.
+
+1. Generate polymorphic annotations (and monomorphic occurrences)
+2. Don't always declare the annotation, so that the annotation can be imported or be declared by another record.
+   The `without declaring` nonterminal modifier causes a declaration to not be generated.
+
+For example:
+
+```silver example
+record Foo without declaring field1 {
+  field1::Integer,
+  field2::String,
+};
+
+record Bar without declaring field2 {
+  field1::Boolean,
+  field2::Float,
+};
+```
+
+The above code should forward to:
+
+```silver example
+nonterminal Foo;
+abstract production foo
+Foo ::=
+{}
+
+annotation field2<a>::a;
+annotation field1<Integer> occurs on Foo;
+annotation field2<String> occurs on Foo;
+
+nonterminal Bar;
+abstract production bar
+Bar ::=
+{}
+
+annotation field1<a>::a;
+annotation field1<Boolean> occurs on Bar;
+annotation field2<Float> occurs on Bar;
+```
+
+Type inference should always be able to sort out what the type of `thing.field1` is correctly, although we don't currently have a way to talk about "types with a `field1` with type `String`."
```

Comment:
You should just be able to write `annotation field1<String> occurs on a` as a type constraint.  Could make an extension forwarding to that too, e.g. `record a has field field1::String`.  Although I really would like to see some more examples of how this would be useful...